### ISSUE TITLE
Let users select a linter (per project) when multiple are available

### ISSUE THREAD

**Issue by AsaAyers** [AtomLinter/linter-plus#31]
`linter-eslint` and `linter-jshint` both claim to lint `source.js`. If the user has both installed they may need to choose to disable one of them.

There might be situations where you do want multiple litners on the same grammar, but I don't have any practical examples.

See https://github.com/AtomLinter/linter-jshint/issues/40#issuecomment-109006308

**Comment by iam4x** [comment]
@AsaAyers I used to have `jscs` and `jshint` together, then `eslint` and `jscs`

It's better to disable it if you don't find the configuration file for the specific linter into project folder

**Comment by steelbrain** [comment]
:-1: We use scopes, and if multiple linters are available for a scope, all of them should be used.
For example, there's one linter that validates the syntax, the other is a handmade linter that finds un-used variables.

and If the user chooses to install both, then the user chose to do so. It's okay for us to lint using them both.

**Comment by iam4x** [comment]
@steelbrain Don't close the discussion, it's a feature was asked on `linter-jshint`, `linter-jscs` and `linter-eslint` has it too.

**Comment by steelbrain** [comment]
Then please provide me with some valid real-life use cases. I am open to ideas. :)
As I mentioned above, if the user installs both the linters, then it's not a use case as the user chose it to be that way.

**Comment by AsaAyers** [comment]
How about 3?

I don't like to use jshint but one project I contribute to uses it. I would
like jshint to be used only for that project.

I'm happy with the defaults of eslint for another project. I shouldn't need
to configure it to make it run.

I'm working on a one off Javascript file. I still want eslint to run while
I'm working on it.

**Comment by steelbrain** [comment]
Then maybe you could look for files like `.jshintrc` in your provider and if that file doesn't exist, then return empty array of errors for that provider. and if it exists then function normally.
Lets not push everything in the core.

**Comment by AsaAyers** [comment]
> I'm happy with the defaults of eslint for another project. I shouldn't need to configure it to make it run.

> I'm working on a one off Javascript file. I still want eslint to run while I'm working on it.

so you consider neither of these to be valid use cases?

**Comment by steelbrain** [comment]
The thing is that I can't think of a decent way to determine which linter to use in the lint function of editor-linter. So that's why I am recommending you to keep this stuff in the linters that have these scenarios.

**Comment by AsaAyers** [comment]
> I can't think of a decent way

discussion over. No need to think about it. everybody go home.

> If the user has both installed they may need to choose to disable one of them.

Linter has detected that you have `linter-eslint` and `linter-jshint`. Which would you prefer to use for this project:
- [x] linter-eslint
- [ ] linter-jshint

and you store that selection along with the path to the root. Next time the user opens that project we just don't activate one of the linters. Maybe they choose to keep both enabled. Probably they won't but it's the user's choice.

We may end up finding we can't or shouldn't do this, but it's worth considering

**Comment by steelbrain** [comment]
Hahahaha, Come On Man! I said I am open to ideas.
So how should the linter know which providers can cause conflicts? Right now we don't even know the names of providers.

**Comment by AsaAyers** [comment]
We know the scopes of all of the linters, so identifying the conflicts is trivial.

Now I remember running into this with another project. You're right that there's no way to get the names short of requiring it as part of the service provider call (bad idea).

Now that we've discussed it, I think this is a decent idea but it isn't actually possible.

**Comment by Eschon** [comment]
How about some kind of optional config file that tells linter to enable/disable specific linters?
Like this:

```
{
  "linter-eslint": "enabled",
  "linter-jshint": "disabled"
}
```

**Comment by steelbrain** [comment]
The question remains the same, how will the linter know the name of the providers, are we gonna add another field to the linter object?

**Comment by AsaAyers** [comment]
I think the best shot would be to see if https://atom.io/packages/project-manager (or something like it) might be able to enable/disable specific plugins per project. Then I'd turn off the whole `linter-jshint` plugin, or disable it by default and only turn it on for one project.

**Comment by steelbrain** [comment]
Agreed AsaAyers

**Comment by Eschon** [comment]
> The question remains the same, how will the linter know the name of the providers, are we gonna add another field to the linter object?

Ok I think I misunderstood the problem.
Is it right that the linter doesn't know by whom it is used and has no way to disable a linter even if it can read it's name through a config file?

**Comment by AsaAyers** [comment]
Every linter plugin provides a [service](https://atom.io/docs/v0.207.0/behind-atom-interacting-with-other-packages-via-services) that the core plugin consumes. There is intentionally no way to know which package is providing the service because the concept is that it doesn't matter. The [original blog post](http://blog.atom.io/2015/03/25/new-services-API.html) (that I skimmed) seems to use the statusbar as an example. so any plugin can provide that service.

So basically we're consuming a bunch of effectively anonymous services that we can't tell appart, so there's no way to know which linter to ignore.
